Add test cases for redirected responses

Rob--W opened the pull request:
These are regression tests for issue #12744 and PR #19028.

Rob--W committed:
Add test cases for redirected responses

Regression tests for issue #12744 and PR #19028

Rob--W commented on test/unit/api_spec.js:
```diff
@@ -2989,17 +2990,14 @@ describe("api", function () {
         const url = new URL(params.url);
-        if (url.hostname === "localhost") {
-          url.hostname = "127.0.0.1";
-        } else if (params.url.hostname === "127.0.0.1") {
```
Note: this `params.url.hostname` was always incorrect, it should have been `url.hostname`.

Rob--W commented:
Note: rebase this when https://github.com/mozilla/pdf.js/pull/19028 is merged. The new test is expected to fail right now because the fix has not been applied yet.

Snuffleupagus commented on test/unit/api_spec.js:
```diff
@@ -2438,12 +2439,12 @@ describe("api", function () {
       The Mozilla Manifesto Addendum
       Pledge for a Healthy Internet
-      
+
```
Please revert this change, since it seems to be causing test failures.

Snuffleupagus commented on test/unit/api_spec.js:
```diff
@@ -2438,12 +2439,12 @@ describe("api", function () {
       It enables new opportunities for learning, building a sense of shared humanity, and solving the pressing problems
       facing people everywhere.
-      
+
```
Please revert this change, since it seems to be causing test failures.

Snuffleupagus commented on test/unit/network_spec.js:
```diff
@@ -116,4 +119,76 @@ describe("network", function () {
+        await rangeReader.read();
+      }
+      rangeReader.cancel(new AbortException("Don't need rangeReader"));
+    }
+    function getCrossOriginUrlWithRedirects({ redirectIfRange = false }) {
```
Nit: Can we please add a blank line here, to aid readability a little bit?
```suggestion
    }

    function getCrossOriginUrlWithRedirects({ redirectIfRange = false }) {
```

Snuffleupagus commented on test/unit/network_spec.js:
```diff
@@ -116,4 +119,76 @@ describe("network", function () {
+        url.searchParams.set("redirectIfRange", "1");
+      }
+      return url.href;
+    }
+    it("redirects allowed if all responses are same-origin", async function () {
```
Nit: Can we please add a blank line here, to aid readability a little bit?
```suggestion
    }

    it("redirects allowed if all responses are same-origin", async function () {
```

Snuffleupagus reviewed (commented):
Thanks for helping out with the tests!

Rob--W commented:
@Snuffleupagus I have addressed the nits, and also updated the PR to run the tests for all IPDFStream implementations where this is relevant, instead of just PDFNetworkStream - i.e. I repeated the test for PDFFetchStream.

Snuffleupagus commented on test/unit/network_spec.js:
```diff
@@ -116,4 +120,88 @@ describe("network", function () {
+    });
+  }
+  describe("Redirects", function () {
+    describe("PDFFetchStream", function () {
+      describeRedirectTests(PDFFetchStream);
+    });
```
Given that `PDFFetchStream` has its tests placed in another file, see https://github.com/mozilla/pdf.js/blob/master/test/unit/fetch_stream_spec.js, it's a bit surprising for these tests to be found here.

Rob--W replied:
Would you prefer me to copy-paste tests across files?

Rob--W replied:
Another option is to put the common test logic in a separate file, e.g. common_pdfstream_tests.js

Let me know what you prefer.

Snuffleupagus replied:
> Another option is to put the common test logic in a separate file, e.g. common_pdfstream_tests.js

That sound like a good idea to me.

Rob--W committed:
Replace createTemporaryNodeServer with TestPdfsServer

Some tests rely on the presence of a server that serves PDF files.
When tests are run from a web browser, the test files and PDF files are
served by the same server (WebServer), but in Node.js that server is not
around.

Currently, the tests that depend on it start a minimal Node.js server
that re-implements part of the functionality from WebServer.

To avoid code duplication when tests depend on more complex behaviors,
this patch replaces createTemporaryNodeServer with the existing
WebServer, wrapped in a new test utility that has the same interface in
Node.js and non-Node.js environments (=TestPdfsServer).

This patch has been tested by running the refactored tests in the
following three configurations:

1. From the browser:
   - http://localhost:8888/test/unit/unit_test.html?spec=api
   - http://localhost:8888/test/unit/unit_test.html?spec=fetch_stream

2. Run specific tests directly with jasmine without legacy bundling:
   `JASMINE_CONFIG_PATH=test/unit/clitests.json ./node_modules/.bin/jasmine --filter='^api|^fetch_stream'`

3. `gulp unittestcli`

Rob--W commented:
@Snuffleupagus I've updated the test by moving it to a separate file. After doing that, I ran into a new issue: while the network_spec.js test ran in the browser only, the fetch_stream_spec.js test was also run in Node.js. I looked into re-using the existing "temporary node server" helper, but then I would have to re-implement the redirect functionality of webserver.mjs there.

So in the end I decided to prepend another commit that replaces `createTemporaryNodeServer` with a new test helper that offers a uniform test server interface. Please let me know what you think of it. I can also split that test server to a separate PR if you'd like.

Snuffleupagus commented:
Locally (on Windows) the new tests unfortunately seem to fail when running `gulp unittest`, which is the "normal" way to run the unit-tests and also what's used on the bots.

/botio unittest

Rob--W commented:
> Locally (on Windows) the new tests unfortunately seem to fail when running `gulp unittest`, which is the "normal" way to run the unit-tests and also what's used on the bots.

Sorry about that, I accidentally lost some lines while rebasing.

This is the diff compared to before:

```diff
diff --git a/test/unit/network_spec.js b/test/unit/network_spec.js
index 4c2624a8b..367e52560 100644
--- a/test/unit/network_spec.js
+++ b/test/unit/network_spec.js
@@ -16,6 +16,7 @@
 import { AbortException } from "../../src/shared/util.js";
 import { PDFNetworkStream } from "../../src/display/network.js";
 import { testCrossOriginRedirects } from "./common_pdfstream_tests.js";
+import { TestPdfsServer } from "./test_utils.js";
 
 describe("network", function () {
   const pdf1 = new URL("../pdfs/tracemonkey.pdf", window.location).href;
@@ -118,6 +119,14 @@ describe("network", function () {
   });
 
   describe("Redirects", function () {
+    beforeAll(async function () {
+      await TestPdfsServer.ensureStarted();
+    });
+
+    afterAll(async function () {
+      await TestPdfsServer.ensureStopped();
+    });
+
     it("redirects allowed if all responses are same-origin", async function () {
       await testCrossOriginRedirects({
         PDFStreamClass: PDFNetworkStream,

```

Rob--W commented:
/botio unittest

Rob--W commented:
Oddly, the test passes in Firefox but failed in Chrome, with the following error:

> TEST-UNEXPECTED-FAIL | redirects blocked if any response is cross-origin | in chrome | Expected a promise to be rejected but it was resolved. Expected a promise to be rejected but it was resolved.

I updated the patch to add a cache buster, to see if that fixes the test.

/botio unittest

Rob--W commented:
??? The test output is inconsistent. Within the same file in the test on Linux, I see the following (one for network, one for fetch_stream, unclear which is which): http://54.241.84.105:8877/efefda0bf24c87e/output.txt

> TEST-UNEXPECTED-FAIL | redirects blocked if any response is cross-origin | in chrome | Expected a promise to be rejected but it was resolved. Expected a promise to be rejected but it was resolved. 
TEST-PASSED | redirects blocked if any response is cross-origin | in chrome

Rob--W commented:
Another attempt - adding Vary:Origin at the server side instead of cache buster at the client side.

Rob--W commented:
/botio unittest

Rob--W commented:
I suspect that Chrome serves the range response directly from its own cache, without reaching the server, and thus never triggering the redirect.

I'll look into a way to avoid that, later.

Rob--W commented:
I've added `Cache-Control: no-store,max-age=0` to the relevant requests from the new test.

/botio unittest

Rob--W commented:
@Snuffleupagus The tests now pass in Chrome, it looks like my analysis of the issue along with the fix is correct. Please review again.

Analysis:

> I suspect that Chrome serves the range response directly from its own cache (based on the response from the full request), without reaching the server, and thus never triggering the redirect, and therefore not triggering the expected failure.
> 
> I'll look into a way to avoid that, later.

Fix:

> I've added `Cache-Control: no-store,max-age=0` to the relevant requests from the new test.

Snuffleupagus commented on test/unit/common_pdfstream_tests.js:
```diff
@@ -0,0 +1,98 @@
+  expect(fullReader.isRangeSupported).toEqual(true);
+
+  // When range requests are supported, the initial full request is aborted.
+  if (PDFStreamClass.name === "PDFNetworkStream") {
+    expect(stream._manager.pendingRequests).toEqual({});
+    expect(fullReader._done).toEqual(false);
+  } else if (PDFStreamClass.name === "PDFFetchStream") {
+    expect(fullReader._abortController.signal.aborted).toEqual(true);
```
Reaching into the class-instances and checking various semi-private properties, that we may at some point want to convert to actually private fields, feels a little bit undesirable to me.

Rob--W replied:
This check is not critical. If we ever refactor this, we can either add a "public" interface to look up the state of the request, or simply drop the test logic here.

The current implementations abort when the server signals support for range requests, but I didn't see test coverage for that, so I added some minimal checks here.

If you prefer to not inspect internal members at the expense of test coverage, I can drop these checks from the test.

Snuffleupagus replied:
I think that we should do *either* of the following, whatever is easiest here:
 - Just drop the checks of "internal" properties.
 - Add test-only helpers, protected by pre-processor checks [similar to here](https://github.com/mozilla/pdf.js/blob/22babd722fcdc6dddd9d77bf51a3eb39f58ff8b2/src/display/api.js#L802-L819), such that we can call one method regardless of what `IPDFStream` implementation we're testing.

Rob--W replied:
I dropped the checks, as they are not critical to this test. Here is the diff between the previous and current version of the patch:

```diff
diff --git a/test/unit/common_pdfstream_tests.js b/test/unit/common_pdfstream_tests.js
index f3efb4348..76ad87b4d 100644
--- a/test/unit/common_pdfstream_tests.js
+++ b/test/unit/common_pdfstream_tests.js
@@ -42,16 +42,8 @@ async function testCrossOriginRedirects({
   await fullReader.headersReady;
   // Sanity check: We can only test range requests if supported:
   expect(fullReader.isRangeSupported).toEqual(true);
-
-  // When range requests are supported, the initial full request is aborted.
-  if (PDFStreamClass.name === "PDFNetworkStream") {
-    expect(stream._manager.pendingRequests).toEqual({});
-    expect(fullReader._done).toEqual(false);
-  } else if (PDFStreamClass.name === "PDFFetchStream") {
-    expect(fullReader._abortController.signal.aborted).toEqual(true);
-  } else {
-    throw new Error(`Unexpected stream class: ${PDFStreamClass.name}`);
-  }
+  // ^ When range requests are supported, the initial full request is aborted,
+  // and we do not need to call fullReader.cancel().
 
   const rangeReader = stream.getRangeReader(
     basicApiFileLength - rangeSize,
```

Snuffleupagus commented on test/unit/test_utils.js:
```diff
@@ -122,73 +122,91 @@ function createIdFactory(pageIndex) {
+  static #startCount = 0;
+
+  static async ensureStarted() {
+    if (this.#startCount++) {
+      // Already started before. E.g. from another beforeAll call.
+      return;
```
If there's two back-to-back calls to `ensureStarted` then the second one may resolve *before* the server is actually available, which seems like it could cause trouble, since loading/initializing is async below.

Snuffleupagus commented on test/unit/test_utils.js:
```diff
@@ -122,73 +122,91 @@ function createIdFactory(pageIndex) {
+    }
+
+    await new Promise(resolve => {
+      this.#webServer.stop(resolve);
```
Similar to the comment above, (theoretically) it might be possible to reach this line *before* the server is actually available.

Snuffleupagus commented on test/unit/common_pdfstream_tests.js:
```diff
@@ -0,0 +1,90 @@
+  await fullReader.headersReady;
+  // Sanity check: We can only test range requests if supported:
+  expect(fullReader.isRangeSupported).toEqual(true);
+  // ^ When range requests are supported, the initial full request is aborted,
+  // and we do not need to call fullReader.cancel().
```
```suggestion
  // ^ When range requests are supported *and* streaming is disabled,
  // the initial full request is aborted and we do not need to call fullReader.cancel().
```

Snuffleupagus reviewed (approved):
r=me, assuming passing unit-tests; thank you.

Snuffleupagus commented:
/botio unittest
